controller/src/main/java/io/pravega/controller/task/Stream/StreamTransactionMetadataTasks.java

```diff
@@ -171,31 +165,35 @@ public void awaitInitialization() throws InterruptedException {
      * @param clientFactory Client factory reference.
      * @param config Controller event processor configuration.
      */
+    @Synchronized
     public Void initializeStreamWriters(final EventStreamClientFactory clientFactory,
                                         final ControllerEventProcessorConfig config) {
-        this.commitStreamName = config.getCommitStreamName();
-        this.commitEventEventStreamWriter = clientFactory.createEventWriter(
-                config.getCommitStreamName(),
-                ControllerEventProcessors.COMMIT_EVENT_SERIALIZER,
-                EventWriterConfig.builder().build());
-
-        this.abortStreamName = config.getAbortStreamName();
-        this.abortEventEventStreamWriter = clientFactory.createEventWriter(
-                config.getAbortStreamName(),
-                ControllerEventProcessors.ABORT_EVENT_SERIALIZER,
-                EventWriterConfig.builder().build());
-
+        if (!commitWriterFuture.isDone()) {
+            commitWriterFuture.complete(clientFactory.createEventWriter(
+                    config.getCommitStreamName(),
+                    ControllerEventProcessors.COMMIT_EVENT_SERIALIZER,
+                    EventWriterConfig.builder().build()));
+        }
+        if (!abortWriterFuture.isDone()) {
+            abortWriterFuture.complete(clientFactory.createEventWriter(
+                    config.getAbortStreamName(),
+                    ControllerEventProcessors.ABORT_EVENT_SERIALIZER,
+                    EventWriterConfig.builder().build()));
+        }
         this.setReady();
         return null;
     }
 
     @VisibleForTesting
+    @Synchronized
     public Void initializeStreamWriters(final String commitStreamName, final EventStreamWriter<CommitEvent> commitWriter,
                                         final String abortStreamName, final EventStreamWriter<AbortEvent> abortWriter) {
-        this.commitStreamName = commitStreamName;
-        this.commitEventEventStreamWriter = commitWriter;
-        this.abortStreamName = abortStreamName;
-        this.abortEventEventStreamWriter = abortWriter;
+        if (!commitWriterFuture.isDone()) {
```

Comment:
I believe that you don't need to check if `isDone()` returns true. If you call `complete` it will ignore your request if the future is already completed (normally or exceptionally).

Reply:
Same above.

Reply:
the reason is, we create the new writer unnecessarily if it is already done. 
isDone check avoids that.